overlord/fdestate/fdestate.go

```diff
@@ -0,0 +1,305 @@
+}
+
+type KeyslotRoleInfo struct {
+	// PrimaryKeyId is the ID for the primary key found in
+	// PrimaryKeys field of FdeState
+	PrimaryKeyId int `json:"primary-key-id"`
+	// Parameters is indexed by container role name
+	Parameters map[string]KeyroleSlotParameters `json:"params,omitempty"`
+	// Tpm2PCRPolicyRevocationCounter is the handle for the TPM
+	// policy revocation counter.  A value of 0 means it is not
+	// set.
+	Tpm2PCRPolicyRevocationCounter uint32 `json:"tpm2-pcr-policy-revocation-counter,omitempty"`
+}
+
+type hashAlg crypto.Hash
+
+func (h *hashAlg) UnmarshalJSON(b []byte) error {
+	var s string
+	if err := json.Unmarshal(b, &s); err != nil {
+		return err
+	}
+	switch s {
+	case "sha256":
+		*h = hashAlg(crypto.SHA256)
+	default:
+		return fmt.Errorf("Unknown algorithm %s", s)
+	}
+
+	return nil
+}
+
+func (h hashAlg) MarshalJSON() ([]byte, error) {
+	switch crypto.Hash(h) {
+	case crypto.SHA256:
+		return json.Marshal("sha256")
+	default:
+		return nil, fmt.Errorf("Unknown algorithm %v", h)
+	}
+}
+
+type KeyDigest struct {
+	// Algorithm is the algorithm for
+	Algorithm hashAlg `json:"alg"`
+	// Salt is the salt for the HMAC digest
+	Salt []byte `json:"salt"`
+	// Digest is the result of `HMAC(key, salt)`
+	Digest []byte `json:"digest"`
+}
+
+const defaultHashAlg = crypto.SHA256
+
+func getPrimaryKeyDigest(devicePath string) (KeyDigest, error) {
+	salt, digest, err := secbootGetPrimaryKeyHash(devicePath, crypto.Hash(defaultHashAlg))
+	if err != nil {
+		return KeyDigest{}, err
+	}
+
+	return KeyDigest{
+		Algorithm: hashAlg(defaultHashAlg),
+		Salt:      salt,
+		Digest:    digest,
+	}, nil
+}
+
+func (kd *KeyDigest) verifyPrimaryKeyDigest(devicePath string) (bool, error) {
+	return secbootVerifyPrimaryKeyHash(devicePath, crypto.Hash(kd.Algorithm), kd.Salt, kd.Digest)
+}
+
+type PrimaryKeyInfo struct {
+	Digest KeyDigest `json:"digest"`
+}
+
+type FdeState struct {
+	// PrimaryKeys are the keys on the system. Key with ID 0 is
+	// reserved for snapd and is populated on first boot. Other
+	// IDs are for externally managed keys.
+	PrimaryKeys map[int]PrimaryKeyInfo `json:"primary-keys"`
+
+	// KeyslotRoles are all keyslot roles indexed by the role name
+	KeyslotRoles map[string]KeyslotRoleInfo `json:"keyslot-roles"`
+}
+
+const fdeStateKey = "fde"
+
+func ensureState(st *state.State) error {
+	var s FdeState
+	err := st.Get(fdeStateKey, &s)
+	if err == nil {
+		// TODO: Do we need to do something in recover?
+		return nil
+	}
+
+	if !errors.Is(err, state.ErrNoState) {
+		return err
+	}
+
+	dataUUID, dataErr := disksDmCryptUUIDFromMountPoint(dirs.GlobalRootDir)
+	saveUUID, saveErr := disksDmCryptUUIDFromMountPoint(dirs.SnapSaveDir)
+	if errors.Is(saveErr, &disks.ErrMountPointNotFound{}) {
+		// TODO: do we need to care about old cases where there is no save partition?
+		return nil
+	}
+
+	if errors.Is(dataErr, disks.ErrNoDmUUID) && errors.Is(saveErr, disks.ErrNoDmUUID) {
+		return nil
+	}
```

Comment:
I think we should look at the sealing method with `device.SealedKeysMethod` instead.

Reply:
I will make it a TODO for now.